### PR TITLE
Feature/permissions

### PR DESCRIPTION
**Pull request by bryan-robitaille**
# Summary | Résumé

This PR modifies the application's access control from a Role Based architecture to an Asset Based architecture.
 - It creates multiple base rules that define user interaction with Asset Objects
 - adds the user's abilities to the authenticated session
 - verifies a user's privileges when accessing pages, and assets, and performing actions on those assets.
 - removes the UserRoles of Administrator and Program Administrator
 - provides a clean database migration from the previous Role Based architecture to the new Asset Based architecture
 - associates Users with ownership of Forms
 - renames the previous Form User to ApiUser
 -  ensures permissions are applied in the correct order by applying 'priorities' to permissions sets.

Implements a new User Management Interface:

| Before                                          | After                                        |
| ----------------------------------------------- | -------------------------------------------- |
| Screen Shot 2022-10-24 at 11 54 54 AM | Screen Shot 2022-10-24 at 11 49 03 AM |
| Screen Does not Exist | Screen Shot 2022-10-24 at 11 49 11 AM|

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [x] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [x] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [ ] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: pages/admin/privileges.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
+import { serverSideTranslations } from "next-i18next/serverSideTranslations";
+import { requireAuthentication } from "@lib/auth";
+
```
Comment (Moro-Code) [inline pages/admin/privileges.tsx:3]: remove space

File: pages/admin/vault.tsx [hunk 2]
```diff
@@ -1,14 +1,13 @@
 import React, { useState, useEffect } from "react";
 import { useTranslation } from "next-i18next";
 import axios from "axios";
-import { serverSideTranslations } from "next-i18next/serverSideTranslations";
+// import { serverSideTranslations } from "next-i18next/serverSideTranslations";
```
Comment (Moro-Code) [inline pages/admin/vault.tsx:4]: Why this change ? Doesn't really matter since this page is not really used. Honestly if you want to make changes here you might as well just remove the page

File: package.json [hunk 3]
```diff
@@ -58,6 +60,7 @@
     "jsonwebtoken": "8.5.1",
     "lexical": "^0.5.0",
     "lint-staged": "13.0.3",
+    "lodash": "4.17.21",
```
Comment (Moro-Code) [inline package.json:63]: lodash is a pretty large dependency. We should follow the existing example in the package.json and only install what we need https://www.npmjs.com/package/lodash.get

**Review by Moro-Code — CHANGES_REQUESTED** [review]
First off I want to start off by saying super super fantastic work. The amount of changes made my head spin honestly and I can't imagine it was easy to have to jump between contexts in the code base. I tried my best to be thorough in my review. I still probably missed some things. There is a lot of comments and obviously we have a limited amount of time to deliver this feature. Comments you feel can be addressed at a later date we can create issues for and resolve them that way. I'm happy to pair program on things or go over things in a meeting if there is any confusion. Fantastic work once again !

File: package.json (hunk 3 above)
Comment (craigzour) [reply package.json:63]: In the code where I used the `get` I am only importing that specific function from Lodash. Is not it just loading what is required thus not the whole library?

Comment (craigzour) [reply package.json:63]: Or maybe you are just worried about the size of our client package at the end?

Comment (Moro-Code) [reply package.json:63]: Yes worried about the size of the client package at the end . Im not sure if webpack would bundle lodash in its entirety

File: components/globals/Base.js [hunk 4]
```diff
@@ -8,13 +8,12 @@ import Fip from "./Fip";
...
 import { useTranslation } from "next-i18next";
 import { getPageClassNames } from "@lib/routeUtils";
-import { UserRole } from "@prisma/client";
 
 const Base = ({ children }) => {
   const formRecord = children?.props?.formRecord;
   const classes = getPageClassNames(formRecord);
 
-  const isAdmin = children?.props?.user?.role === UserRole.ADMINISTRATOR;
+  const isAuthenticated = children?.props?.user;
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply components/globals/Base.js:16]: Modified variable to be a boolean

File: lib/hooks/useAccessControl.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import { PureAbility, Abilities } from "@casl/ability";
+
+interface AccessControlInterface {
+  ability: PureAbility<Abilities, unknown> | null;
+  refreshAbility: () => Promise<void>;
+}
+const AbilityContext = createContext<AccessControlInterface>({} as AccessControlInterface);
+
+export const AccessControlProvider = ({ children }: { children: React.ReactNode }) => {
+  // use a memo function here to set the default for abilitiy
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/hooks/useAccessControl.tsx:13]: No longer relevant, removing.

File: lib/auth.ts [hunk 6]
```diff
@@ -33,60 +36,67 @@ export enum LoggingAction {
...
-          destination: `/${context.locale}${loginUrl(authRole)}`,
-          permanent: false,
-        },
-      };
-    }
+      if (!session) {
+        // If no user, redirect to login
+        return {
+          redirect: {
+            destination: `/${context.locale}/admin/login`,
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/auth.ts:52]: Changed to redirect to cognito login.

File: lib/middleware/middleware.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@ export const middleware = (
     try {
       let props = {};
       for (const middlewareLayer of middlewareArray) {
-        const { next: pass, props: middlewareProps } = await middlewareLayer(req, res, props);
+        const { next: pass, props: middlewareProps } = await middlewareLayer(req, res);
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/middleware/middleware.ts:19]: None of the existing middleware functions were leveraging incoming props.  In an effort to reduce the possiblity of having order dependant middleware layers the `props` argument to `middlewareLayer` was removed.  In the future we see there's a need to have dependenceis between middlewareLayer's it's a feature we could look at solidifying.

File: lib/privileges.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+  }
+  const result = rules.map(({ action, subject }) => {
+    if (isForceTyping(subject)) {
+      return ability.can(action, setSubjectType(subject.type, subject.object));
+    } else {
+      return ability.can(action, subject);
+    }
+  });
+  let accessAllowed = false;
+  if (process.env.NODE_ENV !== "production") {
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/privileges.ts:227]: Pino does have the checking in place to only log the debug level in not prod and info level in prod.  What this environment check if for is to avoid the app going through the map function in prod only to not log the debug strings.  Slight optimization for the production environment only.

File: pages/api/auth/[...nextauth].ts [hunk 9]
```diff
@@ -113,8 +114,8 @@ export const authOptions: NextAuthOptions = {
         token.userId = user.id;
         token.authorizedForm = null;
         token.lastLoginTime = new Date();
-        token.role = user.role;
         token.acceptableUse = false;
+        token.privileges = user.privileges.map((privilege) => privilege.permissions);
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/api/auth/[...nextauth].ts:118]: Good catch.  It was previously but we now have a dedicated function in the session callback that handles it.

File: prisma/schema.prisma [hunk 10]
```diff
@@ -69,19 +69,32 @@ model Template {
...
+}
+
+model Privilege {
+  id          String  @id @default(cuid())
+  nameEn        String @unique
+  nameFr        String @unique
+  descriptionEn String?
+  descriptionFr String?
+  permissions Json
+  priority    Int
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply prisma/schema.prisma:83]: We can have multiple Privileges with the same priority.  The priority is only to ensure that Privileges that have a higher priority are actioned frist.  So if multiple permissions have the priority of '2' they are independant and have to relation to other.  In the migration I listed them 1 - 7 for simplicities sake.

File: prisma/migrations/20221006181804_permissions/migration.sql [hunk 11]
```diff
@@ -0,0 +1,86 @@
+/*
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply prisma/migrations/20221006181804_permissions/migration.sql:1]: Merged migrations into a single migration.

File: prisma/migrations/20221006181804_permissions/migration.sql [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ALTER TABLE "_PrivilegeToUser" ADD CONSTRAINT "_PrivilegeToUser_A_fkey" FOREIGN KEY ("A") REFERENCES "Privilege"("id") ON DELETE CASCADE ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "_PrivilegeToUser" ADD CONSTRAINT "_PrivilegeToUser_B_fkey" FOREIGN KEY ("B") REFERENCES "User"("id") ON DELETE CASCADE ON UPDATE CASCADE;
+
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+-- Populate with Privileges
+
+INSERT INTO
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply prisma/migrations/20221006181804_permissions/migration.sql:54]: I took on that challenge and created a new seed file that handles seeding both production and development environments.  The Inserts have been removed from the sql file.

File: package.json (hunk 3 above)
Comment (bryan-robitaille) [reply package.json:63]: removed lodash and replaced with lodash.get

File: pages/api/users.ts [hunk 13]
```diff
@@ -1,48 +1,56 @@
...
+    res.status(200).json([...users]);
   }
 };
 
-const updateUserAdminStatus = async (
+const updatePrivilegeOnUser = async (
+  ability: Ability,
   req: NextApiRequest,
   res: NextApiResponse,
   session?: Session
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/api/users.ts:27]: :+1

File: pages/api/users.ts [hunk 14]
```diff
@@ -51,13 +59,22 @@ const handler = async (
...
 ): Promise<void> => {
-  switch (req.method) {
-    case "PUT":
-      await updateUserAdminStatus(req, res, session);
-      break;
-    case "GET":
-      await getUserAdminStatus(res);
-      break;
+  try {
+    if (!session) return res.status(401).json({ error: "Unauthorized" });
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/api/users.ts:63]: On the type MiddlewareProps session may or may not exist depending on which middleware is run.  I agree that we know middleware runs and session will indeed be there but TypeScript will still infer that `session` can be undefined given it's optional status on MiddlewareProps.
If ever the sessionExists middleware is removed or modified by so that session is not included the check will ensure a graceful error and return instead of throwing a Type Error.

File: pages/_app.tsx [hunk 15]
```diff
@@ -39,22 +39,25 @@ const MyApp: React.FC<AppPropsWithLayout> = ({
   return (
     <SessionProvider
       session={session}
-      // Re-fetch session every 30 minutes
-      refetchInterval={30 * 60}
+      // Re-fetch session every 5 minutes
+      refetchInterval={5 * 60}
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/_app.tsx:43]: Good catch.  This was legacy code from before we were leveraging cache and callbacks.

File: pages/admin/users.tsx [hunk 16]
```diff
@@ -1,120 +1,208 @@
...
 
-interface UserProps {
-  users: AuthenticatedUser[];
-}
+  return (
+    <div>
+      <div className="mb-8">{`${t("managePermissionsFor")} ${user.name}`}</div>
+      <ul className="flex flex-row flex-wrap gap-8 pb-8 pl-0 list-none">
+        {privileges?.map((privilege) => {
+          const active = userPrivileges.includes(privilege.id);
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/admin/users.tsx:73]: Future Refactor.  No designs exist for this screen, it's just a maintenance screen for now.

File: lib/privilegeCache.ts [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    logMessage.error(e as Error);
+    throw new Error("Could not connect to cache");
+  }
+};
+
+/*
+  Privileges
+*/
+
+export const privilegeCheck = async (userID: string): Promise<Permission[] | null> => {
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/privilegeCache.ts:60]: Originally the plan was we would have been one generic cache funtion module that could have been leveraged from different modules but it seems everytime we added caching funtionality in the app we just copied the generic module and modified it.  Will refactor file to remove uncessary generic functions.

File: lib/privilegeCache.ts [hunk 18]
```diff
@@ -0,0 +1,99 @@
+import { logMessage } from "@lib/logger";
+import { getRedisInstance } from "./integration/redisConnector";
+import { Permission } from "./policyBuilder";
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/privilegeCache.ts]: Refactored caching to move files to cache folder.

File: lib/templates.ts [hunk 19]
```diff
@@ -94,17 +60,117 @@ async function _createTemplate(config: BetterOmit<FormRecord, "id">): Promise<Fo
...
+            },
+          },
+        }),
+      },
+      select: {
+        id: true,
+        jsonConfig: true,
+      },
+    })
+    .catch((e) => prismaErrors(e, []));
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/templates.ts:97]: As we chatted about the other day it's a little more complex then try / catch vs .catch.  The prismaErrors function is an Error filter that will rethrow the error if it doesn't meet specific conditions.  Placing it in a try / catch block wouldn't fit all of the patterns we have accross the app. There's few ways we can do this and for now I think we're just testing out this pattern.  The other pattern would be to have `prismaErrors` in a try catch block but we would need to ensure it is the the absolute last returned line in a try / catch other wise if we are trying to test for other Error types in the catch block they wouldn't be hit.
We can have a larger discussion around this in the future.

File: lib/users.ts [hunk 20]
```diff
@@ -1,128 +1,129 @@
...
+        select: {
+          id: true,
+          name: true,
+          email: true,
+          privileges: true,
+        },
+      })
+      .catch((e) => prismaErrors(e, null));
+  } else {
+    return await prisma.user
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/users.ts:66]: I'll add in an warn so we can trace if something has gone wrong but my initial thoughts around this code is that if a user does become disconneted from privileges either through malicious or accidental means the user will always be reconnected to the absolute base privileges in the app.  It just allivates a support ticket, ensures the user can continue working, and we can investigate later.

File: pages/admin/privileges.tsx [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+}) => {
+  const { ability } = useAccessControl();
+  const { i18n } = useTranslation();
+
+  return (
+    <tr className="border-b-1">
+      <td>{i18n.language === "en" ? privilege.nameEn : privilege.nameFr}</td>
+      <td>{i18n.language === "en" ? privilege.descriptionEn : privilege.descriptionFr}</td>
+      <td>
+        <table className="table-fixed min-w-full text-center">
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/admin/privileges.tsx:30]: This is purely a break glass in case of emergency interface.  Will definitely be completely redesigned and rebuilt.

File: lib/privileges.ts [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+
+    const response = await prisma.privilege.create({
+      data: privilege,
+
+      select: {
+        id: true,
+      },
+    });
+    // Flush existing privilege cache for all users asynchronously
+    flushValues();
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/privileges.ts:230]: No, because the `flushValues()` can be a long running event.  It is not critical that the function resolves before returning.  If the fn rejects the expiry keys on entries will handle the flush over a period up to 10 min.  After 10 min we're guarenteed that the cache is completely refreshed.

File: lib/privileges.ts [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+export const updatePrivilegesForUser = async (
+  ability: Ability,
+  userID: string,
+  privileges: { id: string; action: "add" | "remove" }[]
+) => {
+  try {
+    checkPrivileges(ability, [{ action: "update", subject: "User" }]);
+    const addPrivileges: { id: string }[] = [];
+    const removePrivileges: { id: string }[] = [];
+    privileges.forEach((privilege) => {
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply lib/privileges.ts:131]: lol I would argue that depends on the reader.

File: pages/_app.tsx [hunk 24]
```diff
@@ -39,22 +39,25 @@ const MyApp: React.FC<AppPropsWithLayout> = ({
...
+          )}
+        </SafeHydrate>
+      </AccessControlProvider>
     </SessionProvider>
   );
 };
 
-export default appWithTranslation(MyApp, i18nextConfig);
+// Temporary workaround until this issue is resolved with next-i18next (https://github.com/i18next/next-i18next/issues/1944)
+export default appWithTranslation(MyApp as React.FC);
```
[the remark this answers is not in the document]
Comment (bryan-robitaille) [reply pages/_app.tsx:63]: updated versions

**Review by Moro-Code — APPROVED** [review]
Looks good to go now ! Amazing work
